core/shared/src/main/scala/fs2/concurrent/Channel.scala

```diff
@@ -116,174 +118,179 @@ sealed trait Channel[F[_], A] {
+                  else
+                    FalseF
+                }
+              } else {
+                FalseF
+              }
             }
-            .flatten
-            .uncancelable
-
-        def isClosed = closedGate.tryGet.map(_.isDefined)
 
-        def closed = closedGate.get
-
-        def stream = consumeLoop.stream
-
-        def consumeLoop: Pull[F, A, Unit] =
-          Pull.eval {
-            F.deferred[Unit].flatMap { waiting =>
-              state
-                .modify { state =>
-                  if (shouldEmit(state)) (empty(state.closed), state)
-                  else (state.copy(waiting = waiting.some), state)
-                }
-                .flatMap {
-                  case s @ State(
-                        initValues,
-                        stateSize,
-                        ignorePreviousWaiting @ _,
-                        producers,
-                        closed
-                      ) =>
-                    if (shouldEmit(s)) {
-                      var size = stateSize
-                      val tailValues = List.newBuilder[A]
-                      var unblock = F.unit
-
-                      producers.foreach { case (value, producer) =>
-                        size += 1
-                        tailValues += value
-                        unblock = unblock <* producer.complete(())
+          def send(a: A): F[Either[Channel.Closed, Unit]] = {
+            // we track the outstanding blocked offers so we can distinguish closing from draining
+            // the very last blocked send, when closed, is responsible for triggering the sentinel
+            def permit[E](fe: F[E]): F[E] =
+              MonadCancel[F].uncancelable { poll =>
+                (leasesR.update(_ + 1) *> poll(fe)).guarantee {
+                  leasesR.updateAndGet(_ - 1).flatMap { leases =>
+                    if (leases <= 0) {
+                      closedR.get.flatMap { b =>
+                        if (b)
+                          q.offer(Sentinel)
+                            .start
+                            .void // we don't want to backpressure on processing the sentinel
+                        else
+                          Applicative[F].unit
                       }
-
-                      val toEmit = makeChunk(initValues, tailValues.result(), size)
-
-                      unblock.as(Pull.output(toEmit) >> consumeLoop)
                     } else {
-                      F.pure(
-                        if (closed) Pull.done
-                        else Pull.eval(waiting.get) >> consumeLoop
-                      )
+                      Applicative[F].unit
                     }
+                  }
                 }
-                .uncancelable
-            }
-          }.flatten
+              }
 
-        def notifyStream(waitForChanges: Option[Deferred[F, Unit]]) =
-          waitForChanges.traverse(_.complete(()))
+            isClosed.ifM(
```

Comment:
This whole function I feel is pretty complex, _and_ has a ton of access to `closed`, just because we split the Refs

Reply:
The double-check is actually fundamental. Ultimately this stems from the fact that we need atomicity between `leases` and `q`, which are split by definition.

Reply:
Mm, it doesn't strike me as fundamental at all because I don't think it's about atomicity between `leases` and `q` (which you can't get anyway):
The first check is there because you don't want to increase the number of leases if the queue is already closed (to ensure that closing logic is done by `close`), you then increase the number of leases, but have to check again to make sure the queue hasn't been closed in the meantime, so you can avoid offering an element in that case. So the issue is the lack of atomicity between `leases` and `closed`, which is my whole point: if they were in one Ref, there couldn't be a case where you increase the leases but the queue has been closed, and you wouldn't need the double check.

Reply:
Kinda off the top of my head, but this does seem simpler to me:
```scala
def send(a: A): F[Either[Channel.Closed, Unit]] = {
  val doSend = q.offer(a.asInstanceOf[AnyRef]).as(RightUnit)
  val cleanup = state.modify { case State(leases, closed) =>
    (
      State(leases - 1, closed),
      q.offer(Sentinel).start.void.whenA(leases == 1 && closed)
    )
  }.flatten


  F.uncancelable { poll =>
    state.modify { case State(leases, closed)  =>
      if (closed) state -> F.unit
      else State(leases + 1, closed) -> poll(doSend).guarantee(cleanup)
    }.flatten
  }
}
```